crates/contract/src/v3_0_2_state.rs

```diff
@@ -40,6 +42,66 @@ impl From<Config> for crate::config::Config {
     }
 }
 
+/// Old version of [`UpdateEntry`] that included votes.
+#[derive(Clone, Debug, BorshSerialize, BorshDeserialize)]
+struct UpdateEntry {
+    update: Update,
+    votes: HashSet<AccountId>,
+    bytes_used: u128,
+}
+
+/// Old version of [`ProposedUpdates`] that used the old [`UpdateEntry`] with votes.
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub struct ProposedUpdates {
+    vote_by_participant: IterableMap<AccountId, UpdateId>,
+    entries: IterableMap<UpdateId, UpdateEntry>,
+    id: UpdateId,
+}
+
+impl From<ProposedUpdates> for crate::update::ProposedUpdates {
+    fn from(old: ProposedUpdates) -> Self {
+        let mut entries =
+            IterableMap::new(crate::storage_keys::StorageKey::ProposedUpdatesEntriesV2);
+
+        // Build map of update_id -> voters for validation (single pass)
+        let mut votes_by_update: BTreeMap<UpdateId, HashSet<AccountId>> = BTreeMap::new();
+        for (account, update_id) in old.vote_by_participant.iter() {
+            votes_by_update
+                .entry(*update_id)
+                .or_default()
+                .insert(account.clone());
+        }
+
+        for (id, entry) in old.entries.iter() {
+            // Validate votes consistency
+            if !votes_by_update
+                .get(id)
+                .map_or(entry.votes.is_empty(), |v| v == &entry.votes)
+            {
+                near_sdk::env::log_str(&format!(
+                    "Migration warning: Inconsistent votes for update {id:?}. Entry votes: {:?}, vote_by_participant: {:?}",
+                    entry.votes,
+                    votes_by_update.get(id)
+                ));
+            }
+
+            entries.insert(
+                *id,
+                crate::update::UpdateEntry {
+                    update: entry.update.clone(),
+                    bytes_used: entry.bytes_used,
+                },
+            );
+        }
+
+        Self {
+            vote_by_participant: old.vote_by_participant,
+            entries,
+            id: old.id,
+        }
+    }
+}
```

Comment:
nit: Do we actually need this logic? I thought once we enter migration, we already cleaned / removed all proposed updates and votes.